barretenberg/cpp/src/barretenberg/benchmark/simulator_bench/simulator.bench.cpp

```diff
@@ -0,0 +1,126 @@
+    using VerificationKey = typename Flavor::VerificationKey;
+    using CircuitSimulator = typename RecursiveFlavor::CircuitBuilder;
+    using SimulatingVerifier = stdlib::recursion::honk::UltraRecursiveVerifier_<RecursiveFlavor>;
+    SimulatorFixture() = default;
+
+    struct VerifierInput {
+        HonkProof proof;
+        std::shared_ptr<typename Flavor::VerificationKey> verification_key;
+    };
+
+    void SetUp([[maybe_unused]] const ::benchmark::State& state) override
+    {
+        bb::srs::init_crs_factory("../srs_db/ignition");
+        bb::srs::init_grumpkin_crs_factory("../srs_db/grumpkin");
+    }
+
+    static VerifierInput create_proof_goblin_builder(bool large = false)
+    {
+
+        auto builder = construct_mock_function_circuit(large);
+        auto instance = std::make_shared<ProverInstance>(builder);
+        UltraProver_<Flavor> prover(instance);
+        auto ultra_proof = prover.construct_proof();
+        auto verification_key = std::make_shared<VerificationKey>(instance->proving_key);
+        return { ultra_proof, verification_key };
+    }
+
+    static Builder construct_mock_function_circuit(bool large = false)
+    {
+        using InnerCurve = bb::stdlib::bn254<Builder>;
+
+        using fr_ct = InnerCurve::ScalarField;
+        using point_ct = InnerCurve::AffineElement;
+        using fr = typename InnerCurve::ScalarFieldNative;
+        using point = typename InnerCurve::GroupNative::affine_element;
+        Builder builder;
+
+        // Add some arbitrary goblin-style ECC op gates via a batch mul
+        size_t num_points = 5;
+        std::vector<point_ct> circuit_points;
+        std::vector<fr_ct> circuit_scalars;
+        for (size_t i = 0; i < num_points; ++i) {
+            circuit_points.push_back(point_ct::from_witness(&builder, point::random_element()));
+            circuit_scalars.push_back(fr_ct::from_witness(&builder, fr::random_element()));
+        }
+        point_ct::batch_mul(circuit_points, circuit_scalars);
+
+        // Determine number of times to execute the below operations that constitute the mock circuit logic. Note
+        // that the circuit size does not scale linearly with number of iterations due to e.g. amortization of lookup
+
+        const size_t NUM_ITERATIONS_LARGE = 12; // results in circuit size 2^19 (502238 gates)
+        const size_t NUM_ITERATIONS_MEDIUM = 3; // results in circuit size 2^17 (124843 gates)
+        const size_t NUM_ITERATIONS = large ? NUM_ITERATIONS_LARGE : NUM_ITERATIONS_MEDIUM;
+
+        stdlib::generate_sha256_test_circuit(builder, NUM_ITERATIONS);             // min gates: ~39k
+        stdlib::generate_ecdsa_verification_test_circuit(builder, NUM_ITERATIONS); // min gates: ~41k
+        stdlib::generate_merkle_membership_test_circuit(builder, NUM_ITERATIONS);  // min gates: ~29k
+
+        return builder;
+    }
+};
+
+BENCHMARK_TEMPLATE_F(SimulatorFixture, GoblinNative, bb::GoblinUltraRecursiveFlavor_<bb::CircuitSimulatorBN254>)
```

Comment:
This and UltraNative are basically the same code (ditto for the simulated ones) but I haven't managed to instantiate with multiple templates, not even sure it's supported. Very much open to suggestions